nextcloud: restrict web server support to nginx; stop sharing nginx user/group; improve setup service

DavHau opened the pull request:
###### Things done
(This description is heavily outdated. Plans changed a lot due to the discussions. See below)
 - user and group used by all nextcloud related services are now unified and configurable. If nginx is enabled, the group will be forced to `nginx` and an error is raised if group is set manually.
- nginx is enabled by default now. (In favor of having a working default configuration)
- phpfpm also uses configured nextcloud user + group
- In the setup service instead of `mkdir` + `chown` now `install` is used to setup the directories.
  This has the disadvantage that the group of nextcloud cannot be changed anymore on an existing installation without having to fix the ownership of files manually. But it think `chown`ing all files on every single startup is bad since it's a very intense operation and can take very long depending on the amount of files stored. Also the `chown` operation itself might not be allowed depending on the filesystem. I for example use a sshfs mount for my nextcloud files where this fails

fixes #69261
fixes #48881

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

DavHau committed:
nextcloud: configurable user and group, enabled nginx, improve setup

aanderse commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -71,6 +73,19 @@ in {
       description = "Which package to use for the Nextcloud instance.";
       relatedPackages = [ "nextcloud17" "nextcloud18" "nextcloud19" ];
     };
+    user = mkOption {
```
What is the purpose of a configurable `user` option?

If the user is configurable then we should follow the pattern of conditionally provisioning users only if the default value is specified:
```
users.users.nextcloud = optionalAttrs (cfg.user == "nextcloud") {
  ...
};
```

DavHau replied:
Actually the user doesn't really need to be configurable. Sharing the group with other services which need to access the same files should be sufficient. So I will revert this change.

aanderse commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -488,38 +504,38 @@ in {
           settings = mapAttrs (name: mkDefault) {
-            "listen.owner" = "nginx";
-            "listen.group" = "nginx";
+            "listen.owner" = cfg.user;
```
This isn't what we want. The `user` option should specify the user that the `nextcloud` service runs as, not the user the web server runs as. `"listen.owner"` is intended to be the user the web server is run as.

DavHau replied:
Then I should change the user back to `nextcloud` (due to the other discussions).
And the `listen.owner/group` should be set to `config.services.nginx.user/group` if nginx is enabled.
But what if nginx is disabled? Should we add optional options to configure the owner/group for the webservice?

aanderse commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -468,11 +486,9 @@ in {
             chmod og+x ${cfg.home}
             ln -sf ${cfg.package}/apps ${cfg.home}/
-            mkdir -p ${cfg.home}/config ${cfg.home}/data ${cfg.home}/store-apps
+            install -o ${cfg.user} -g ${group} -d ${cfg.home}/config ${cfg.home}/data ${cfg.home}/store-apps
```
This script really shouldn't even be run as `root`, so if there are any changes here maybe it should be to fix that...

aanderse commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -71,6 +73,19 @@ in {
+      default = "nextcloud";
+      description = "User of the nextcloud service";
+    };
+    group = mkOption {
```
Another possibility here would be to keep running `nextcloud` services (including `php-fpm`) as a non web server user (like the current `nextcloud` user) and instead add the `nextcloud` group to your web servers `extraGroups` option, like so:
```
mkIf cfg.nginx.enable {
  users.users.nginx.extraGroups = [ cfg.group ]; # in other words "nextcloud"
}
```

The reason for this is that the `nginx` user (and group) likely has access to very sensitive files, such as certificates. If `nginx` can read `nextcloud` files (which you want) that is much less of a risk than having `nextcloud` able to read `nginx` files - a serious security concern. This obviously requires that your `nextcloud` files are `g+r`.

DavHau replied:
Sounds like a good idea. The files created by nextcloud seem to be `g+r`, so this shouldn't be an issue.

DavHau replied:
I wonder how to best change the user/group of nextcloud without breaking existing installations. It would require a one time service run as root which changes the ownership. And I would need to have some state file to save that this setup has been completed.
Alternatively I could leave the default user.group to be nginx if `system.stateVersion <= 20.03`, but i don't really like that approach.

DavHau replied:
Forget about it. Since it looks like the user will still be `nextcloud`, only the group needs to be changed from `nginx` to `nextcloud` which is easy to do in the current setup. Also no state needs to be stored.

aanderse reviewed (commented):
Here are some initial thoughts I had which might be a starting point for further discussion.

Ma27 commented:
Did anyone of you test what happens when deploying this configuration to an existing Nextcloud instance? Also, what's supposed to happen if `cfg.user` changes (or someone switches e.g. from `nginx` to `httpd`?).

DavHau committed:
nextcloud improve user/group handling

- remove optons cfg.user, cfg.groups
- add option `serverUser` which is required when not using nginx
- add `serverUser` to nextcloud group
- set user/group to "nextcloud" for nextcloud services
- make setup-service non-root

DavHau commented:
please check https://github.com/NixOS/nixpkgs/pull/93584/commits/6ee30041327343137bab79cf924eac1a17ad0d5c.
  - removed optons cfg.user, cfg.groups
  - add option `serverUser` which is required when not using nginx
  - add `serverUser` to nextcloud group
  - set user/group to "nextcloud" for nextcloud services
  - make setup-service non-root

Did not yet implement a solution for fixing group ownership of files for existing installations. Your suggestions are welcome.

aanderse commented:
@DavHau for context how are you trying to run `nginx`? `caddy`?

DavHau commented:
Sorry my description of what I changed was not very precise.

All nextcloud services now use the user and group "nextcloud" (hard-coded). The web server user must therefore be part of the the nextcloud group.

The webserver user can be configured via `cfg.serverUser`.
This user will be automatically added to the nextcloud group and it will be used for `listen.owner` of phpfpm. Also its group will automatically be used for `listen.group`

`cfg.serverUser`defaults to the nginx user if nginx is enabled. Otherwise it must be specified manually.

An example configuration using nginx would be:  
(if nothing specified nginx will be enabled and its user added to the nextcloud group)
```nix
{ pkgs, ... }:
{
  networking.firewall.allowedTCPPorts = [80];
  services.nextcloud = {
    enable = true;
    hostName = "10.233.1.2";
    config.adminpass = "test";
    package = pkgs.nextcloud19;
  };
}
```

A configuration for an alternative webserver could  look like this:
```nix
{ pkgs, ... }:
{
  networking.firewall.allowedTCPPorts = [80];
  services.nextcloud = {
    enable = true;
    hostName = "10.233.1.2";
    config.adminpass = "test";
    package = pkgs.nextcloud19;
    
    # disable nginx and set serverUser
    nginx.enable = false;
    serverUser = "caddy";

    # caddy config
    ...
  };
}
```

aanderse commented:
@DavHau by having a configurable web server user in this module it seems we're just circling back around to https://github.com/NixOS/nixpkgs/issues/93070.

@Ma27 I suggest we remove the `nginx.enable` option entirely and just take it as a given: this module is for `nginx` only.

DavHau committed:
nextcloud: restrict web server support to nginx only

DavHau committed:
nextcloud: fix group permissions on startup

DavHau commented:
- removed option `nextcloud.nginx.enable`. Nginx is now the only supported web server.
- modified nextcloud-setup service to fix group ownership of nextcloud files if they don't match.
  All installations prior to this PR will be fixed by this. (changed from nginx -> nextcloud).

aanderse commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -502,136 +501,131 @@ in {
-            "listen.owner" = "nginx";
-            "listen.group" = "nginx";
+            "listen.owner" = config.services.nginx.user;
+            "listen.group" = config.users.users.${config.services.nginx.user}.group;
```
```suggestion
            "listen.group" = config.services.nginx.group;
```

aanderse reviewed (changes requested):
Looking good :+1:

DavHau committed:
nextcloud: shorten nginx group reference

Co-authored-by: Aaron Andersen <aaron@fosslib.net>

aanderse commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -468,10 +458,18 @@ in {
+              if [ ! -e $dir ]; then
+                install -o nextcloud -g nextcloud -d $dir
+              elif [ $(stat -c "%G" $dir) != "nextcloud" ]; then
+                chown -R nextcloud:nextcloud $dir
```
This script is run as `nextcloud` so this `chown` doesn't make sense to me. The code this script replaces also had that, though... can someone explain this to me?

DavHau replied:
The chown command is only for fixing the group permissions for installation prior to this PR.
Those installations will have `nextcloud:nginx` and it will be changed to `nextcloud:nextcloud`. This command is possible as nextcloud user.

Other reasons why I wrote the script like this:
1. We can't just do a plain `install -o nextcloud -g nextcloud -d ${cfg.home}/config ${cfg.home}/data ${cfg.home}/store-apps`. Because if the these dirs already exists with wrong group, it will only fix the group for the top level dirs, but not their contents.
1. We also don't want to do an unconditional `chown -R` on every single startup, since this is an expensive operation and might not even work depending on the fs.

As of this script, it will only create the top level dirs if they don't exist and it will only fix the group if the group is wrong.

aanderse replied:
That makes sense to me. Maybe consider `chgrp` if the intention is to only change the group.

Mic92 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -91,16 +91,6 @@ in {
       '';
     };
 
-    nginx.enable = mkOption {
```
The only argument against removing this option is if your `nginx` is running somewhere else i.e. in a different host/container.

Mic92 replied:
The workaround would be `systemd.services.nginx.enable = false` in this case.

DavHau replied:
You could still either setup a reverse proxy to the nginx inside the container, or disable it via `mkForce services.nginx.enable = false`. Do you think this is acceptable?

Mic92 replied:
I suppose though.

DavHau replied:
But then you would run again into. https://github.com/NixOS/nixpkgs/issues/69261
Therefore it's not really done by switching that one option to false.
I think we need to decide. Either we want to support having nextcloud without nginx (then we need more config options) or we just don't.

DavHau replied:
I would argue that the benefits we get by supporting non-nginx setups aren't big enough to justify the overhead it will have in maintaining it. It just makes things more complicated on every single change. When restricting to nginx in general we can just exclude a lot of question marks whenever we touch anything user/group related
 
Even if you run it inside a container, it shouldn't be a performance bottleneck to have this extra nginx running.

aanderse replied:
Looking for @Ma27 to weigh in on this...

Mic92 replied:
I don't need it personally. We could also wait for someone to add it back ;). However it should get deprecated properly like here: https://github.com/NixOS/nixpkgs/blob/6eadc11005318fc4eef7da19898cff51689fb723/nixos/modules/tasks/network-interfaces.nix#L316

Ma27 replied:
Will think about this tonight @aanderse (oh an please ping me if I forget about this again :sweat_smile:  )

DavHau replied:
Based on the current implementation, i think it would not be super hard to add support for non-nginx. We just need to add another config option to specify the user of the webserver used. This user will be added to the nextcloud group and phpfpm will be opened to that user.

I already had that implemented before but reverted it. I'm happy to re-revert ;).

The downside of it would be that it makes the module harder to understand for people and therefore harder to maintain i guess.

Ma27 replied:
> The downside of it would be that it makes the module harder to understand for people and therefore harder to maintain i guess.

I think I mentioned this in a different thread already, but I might be wrong: I'm rather strongly against supporting multiple reverse-proxies for the exact reasons you gave, even more complexity (and thus more work for the maintainers - including me).

I think that your PR is actually a great solution: we have `nginx` as default (as most of the services here do), but admins now have the freedom to opt-out  of it and use an alternative.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -502,136 +501,131 @@ in {
-              '';
-            };
+    
+      services.nginx.enable = true;
```
What about `mkDefault true` here?

Ma27 reviewed (commented):
The implementation seems fine: `nginx` is in most of the services in our tree that serve HTTP the default choice and now it's possible here to opt-out by disabling it.
Also did a brief test of the altered functionality and it appears fine to me.

However there are a few things I'd like to get resolved before merging this:

* We should mention this in our release-notes. When having a big library in your nextcloud, the `chown` will take some time, so admins should be aware of this *before* deploying the new config.

* Removed options must be documented with `mkRemovedOptionModule`.

* The nextcloud-specific tests don't evaluate because of the removed `nginx.enable` option.

DavHau committed:
nextcloud: deprecate nginx, use chgrp, mkDefault for nginx, fix tests

DavHau commented:
Things done:
 - deprecated `services.nextcloud.nginx.enable` option via mkRemovedOptionModule
 - replaced chown with chgrp
 - fixed tests by removing use of option `services.nextcloud.nginx.enable`
 - use mkDefault for `services.nginx.enable`

> We should mention this in our release-notes. When having a big library in your nextcloud, the chown will take some time, so admins should be aware of this before deploying the new config.

I don't think this will be necessary. Up until this point the setup service always did a chown an all nextcloud files on each startup. Therefore the new implementation doesn't bring any performance drawbacks compared to the old one.

aanderse reviewed (approved):
In general, LGTM :+1: I have not tested and don't use this module, though, so need someone like @Ma27 to approve as well.

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -45,6 +45,12 @@ let
+
+  imports = [
+    ( mkRemovedOptionModule [ "services" "nextcloud" "nginx" "enable" ]
+      "The nextcloud module dropped support for other webservers than nginx.")
```
That's wrong: you can still use anything else (in fact now it's even easier than before!) by setting `services.nginx.enable = false`.

DavHau replied:
That's not entirely true. One also has to change the `listen.owner` and `listen.group` of phpfpm to something else than nginx and also add the user of the new webserver to `users.groups.nextcloud.members`. Should I add an example for disabling nginx to the deprecation text?

>We should mention this in the release notes. The rollout of the new config can take quite long due to the recursive chgrp

As I explained in my last comment, The rollout won't take any longer than the normal startup of the previous implementation. Therefore I think it's unnecessary to state anything in the release notes. Do you agree?

Ma27 replied:
> Should I add an example for disabling nginx to the deprecation text?

Yes, please. But as this is possible (with a few overrides though), I'd also rephrase the message above.

> Therefore I think it's unnecessary to state anything in the release notes. Do you agree?

As mentioned above, I only tested this against a temporary environment. As I'm always slightly anxious when fiddling around with those kinds of scripts (touching user data) in NixOS, I'd prefer to test this against live-data first. I'll hopefully be able to do this tomorrow (with the fat library I'm hosting in my personal Nextcloud :P). After that I think that this should be good to go :)

Ma27 replied:
Tested against my instance, didn't have any problems.

Ma27 reviewed (commented):
LGTM :+1: 

* Did anyone test this with live data yet? I just checked whether the "migration" to the new config works fine in a temporary VM network.
* We should mention this in the release notes. The rollout of the new config can take quite long due to the recursive `chgrp`.

DavHau committed:
nextcloud: use mkDefault for whole nginx config

ajs124 commented:
This broke then manual: [log](https://nix-cache.s3.amazonaws.com/log/sm4ynvc1r7waxjaa8m3vpg450gvv5spi-nixos-manual-combined.drv). Plz fix, kthxbye. Or wait for a 10-12h, then I'll probably get around to it myself.

aszlig commented:
@ajs124: Fixed in 1365b9ac700aabe8dc959e9897bc1376e416c553.

DavHau commented:
Thanks @ajs124 .
@Ma27 I thought we still wanted to improve the phrase `The nextcloud module dropped support for other webservers than nginx.`.
But now the PR has already been merged. Should I make another commit ontop of this PR and repopen it?

Ma27 commented:
@aszlig @ajs124 sorry for that!
@DavHau yeah feel free to do that :) Otherwise I'll take care of this next week.

Ma27 commented:
Just for the record, this PR caused another regression in #95259 which got fixed in fddeb7cb73096786de8e59a8502cf3ae523bd694.

I'm fully aware that I merged this too early which is why I want to openly apologize to everyone who went into trouble because of the (now fixed) regressions I caused with the merge!

GaetanLepage commented:
Hello,

I have read the manual on how to use a different webserver for nextcloud.
However, I don't really get why, to prevent nginx being configured for nextcloud, it is now necessary to **completely** disable nginx.
What if, I want to have the nginx service enabled for something else while still wanting to use, let's say httpd, as the web server for nextcloud ?

Thank you in advance for the clarification :)
